### ISSUE TITLE
[Feature Request] Missing Vorbis? and more

### ISSUE THREAD

**Issue by orbittwz** [DJDoubleD/QobuzDownloaderX-MOD#44]
# **🚀 Feature Request**

## **Is your feature request related to a problem? Please describe.**

* When downloading albums, all the tracks are missing the following tags: CRC tag, Encoder tag.
* When downloading albums, the user cannot choose the folder name of the album, nor it's not one folder instead of many.
* When downloading albums, there are missing Vorbis tags: PUBLISHER, BARCODE.

---

## **Describe the solution you'd like**

* Save all tracks with CRC tag and Encoder tag.
* Adding an option to choose how albums are saved in folders, either by setting a variable of the tags or choose from templates.
* Adding all the missing Vorbis tags.

---

## **Describe alternatives you've considered**

* I don't have any control over this program coding so I can't suggest that.

---

### **Additional context**

* no and thanks.

**Comment by orbittwz** [comment]
Adding some feedback over tracks I got using this version...
All the tracks have this vorbis tags:
![image](https://github.com/DJDoubleD/QobuzDownloaderX-MOD/assets/45605734/db58d0e2-e8e8-427c-80c0-041ee778974b)
This is not needed, and maybe you can omit this or at least have a toggle to keep or not.
Moreover, both PUBLISHER and BARCODE are actually exist in this version already, but not the right vorbis tags.
ORGANIZATION should be PUBLISHER.
UPC should be BARCODE but without the extra zero prefix.
Thanks!

**Comment by DJDoubleD** [comment]
A few remarks regarding this feature request:

- I assume that the CRC requested refers to the MD5 checksum of the unencoded stream content.
This information is not provided by any Qobuz API nor does Qobuz embed it in their source stream. So this info is simply unavailable.
- The same goes for the encoder tag. This is not provided by Qobuz and thus unavailable.
- Regarding changing the album download folder, how does this request differ from the existing requests #24 and/or #36?
- Regarding the ORGANIZATION vs PUBLISHER tags: I referenced the [Vorbis  Comment standard](https://xiph.org/vorbis/doc/v-comment.html) several times the release notes of version 1.2.2.0.
The ORGANIZATION tag is in fact the mentioned standard for the publishing record label, not the PUBLISHER tag. In QBDLX-MOD, just like in the original QBDLX, I use the library [taglib-sharp](https://github.com/mono/taglib-sharp) to write tags to both mp3 files and FLAC files. The library includes the Publisher field as a supported tag and uses the ORGANIZATION as per the specs when writing to FLAC files. And there is in fact a toggle to write this tag to the audio files or not, its the "Label" checkbox.
Also see the discussion in #43 where it's requested to write the label name to the "RECORDLABEL" tag instead...
So before I change this, I will need to see some more convincing documentation that overrules the Vorbis Comment standards that I could find in the linked reference.
- Regarding the ORIGINALRELEASETIME: this will probably be changed when I figure out what to do with #43.
- Regarding the UPC: this is specifically listed as such in the Qobuz API and this tag has not only remained like this for years, it's also used by other platforms (like Deezer) and other download tools. So I'm afraid I will need some definitive documentation regarding this Vorbis Comment tag before changing it.

In conclusion: some requests are simply impossible to provide and other need at least some clarification and/or justification before I can consider to change them. 
It would seem tagging is a delicate subject as everybody has their own preferences so the most objective thing to do, is follow existing standards...

**Comment by orbittwz** [comment]
Thanks for the detailed information...
Do you know if Qobuz API allows further vorbis rewriting with different metadata encoders such as Mutagen or FLAC?

**Comment by DJDoubleD** [comment]
> Thanks for the detailed information... Do you know if Qobuz API allows further vorbis rewriting with different metadata encoders such as Mutagen or FLAC?

I'm afraid I don't understand your question.

The Qobuz API is just an interface to send and receive data from Qobuz. QBDLX-MOD then uses the available received data to write to the most essential and popular tags using [taglib-sharp](https://github.com/mono/taglib-sharp).
Mutagen is just like  [taglib-sharp](https://github.com/mono/taglib-sharp) but in/for python instead of c#.
From their own website : `Mutagen is a Python module to handle audio metadata`.

And FLAC is just a container format with the ability to contain compressed lossless audio and Vorbis Comment tags.

**Comment by orbittwz** [comment]
let me rephrase then:
since Qobuz API doesn't write CRC and Encoder tags, is it possible to do so somehow with your program?
can you use another library to create and insert them into FLAC streams?

**Comment by DJDoubleD** [comment]
> let me rephrase then: since Qobuz API doesn't write CRC and Encoder tags, is it possible to do so somehow with your program? can you use another library to create and insert them into FLAC streams?

Te biggest problem is simply: the source information for the original server side audio stream cannot be retrieved and the same goes for the encoder that was used on to generate the original files.
The only way that I see to generate a MD5 checksum (in the STREAMINFO header, this is not a Vorbis Comment tag!) is to re-encode the downloaded audio files. And that would just be completely unreasonable in my opinion...

**Comment by orbittwz** [comment]
I've never heard of un-Encoder tagged FLACs and this is quite the problem...
Not knowing which Encoder was used to actually compress and encode the audio stream is very weird to say the least!
Can you check if your problem is by some odd way is stripping this kind of information??
Of all my years as audiophile, this is first for me.
Let me know (also do try to check if CRC can be inserted as well to the stream info as you said)
Thanks

**Comment by DJDoubleD** [comment]
> I've never heard of un-Encoder tagged FLACs and this is quite the problem... Not knowing which Encoder was used to actually compress and encode the audio stream is very weird to say the least! Can you check if your problem is by some odd way is stripping this kind of information?? Of all my years as audiophile, this is first for me. Let me know (also do try to check if CRC can be inserted as well to the stream info as you said) Thanks

I agree this is weird, but it's just the way Qobuz serves their FLAC files for streaming.
Maybe they have some kind of custom transcoding workflow which doesn't calculate/write these values to their files, or maybe they use a tool like [ExiffTool](https://exiftool.org/) to strip their files of these 2 fields. I have no idea...

The MD5 checksum of the unencoded stream in a FLAC file is calculated **_during_** the encoding process.
There is no way to (re)calculate this for an already encoded file. You can only transcode the FLAC into a new FLAC, which will automatically add both the encoder (the one used to encode the new file) and MD5 checksum to the FLAC STREAMINFO headers.
This seems redundant to say the least and is way out of scope of a simple downloader...

If I'm wrong, feel free to point me to a documentation link which explains how to recalculate the checksum of an already encoded file without transcoding it and I'll look into it.

**Comment by orbittwz** [comment]
is this the same with bought purchases in site and downloading them?
all the files missing Encoder and CRC tags?

**Comment by orbittwz** [comment]
OK some progress, I've just added the Encoder tag without value using dbpoweramp and the Encoder was added with the track original value...
So basically the downloaded files using your program have the value, not the field??
Can you check if this is possible to fix? same goes to CRC field

**Comment by DJDoubleD** [comment]
> OK some progress, I've just added the Encoder tag without value using dbpoweramp and the Encoder was added with the track original value... So basically the downloaded files using your program have the value, not the field?? Can you check if this is possible to fix? same goes to CRC field

As far as I know, dbpoweramp is a Music Converter. When you add the tag, you literally need to press a button "Convert".
So I highly suspect dbpoweramp simply writes the encoder used by dbpoweramp itself to update the file when you leave the value of the tag empty. 
In your examples, you get a libFLAC build from 2023-06-23 as encoder. I suspect you will get this result for all files you try this manipulation on? 
Please try this on older Qobuz files which must have been encoded on their server before 2023-06-23 and see if you get an older libFLAC build version or the same recent one.
Also, the tagging editor that dbpoweramp uses, seems to be a proprietary tool, so I can't check the source.

**Comment by orbittwz** [comment]
![image](https://github.com/DJDoubleD/QobuzDownloaderX-MOD/assets/45605734/7ee57f60-0c19-455a-b9ad-cd5df2a1744f)
As you can see, the files are from 12/06... so not sure if that manipulation makes it OK?
Did you checked if the program expose the CRC and Encoder fields?

**Comment by DJDoubleD** [comment]
> Seems you are correct, so it's writing the encoder it already have... So it's not a good example...

That's what I was saying all along, the original data simply isn't there unfortunately.

> What can be done so files from Qobuz have their original Encoder and CRC? does downloads using your program different from purchased downloads?

As far as I know, there simply isn't anything to be done to retrieve the original Encoder and CRC.
I can't compare with purchased downloads from Qobuz as I have none (impossible to rip purchase only tracks).
And even if I were to have purchased downloads and they do contain this info, it still wouldn't offer a solution to the missing data in the files they serve up for streaming...

I have exhausted all possible avenues to a solution for your request regarding the encoder and CRC fields, which simply aren't available.
So unless you can provide me with a workable solution, I'll have to close this issue as there is nothing to be done...

**Comment by orbittwz** [comment]
Sure thanks man for your kindness explaining and taking the time answering this...
Will close for now.

**Comment by orbittwz** [comment]
Well I can't be more happy since I just got the answer how to get both values...
Using the FLAC binaries (using 1.4.3) with just -f switch is writing those values properly and not even reencoding the file??
The timestamp hasn't changed but the file now contains both fields!
![image](https://github.com/DJDoubleD/QobuzDownloaderX-MOD/assets/45605734/12393987-1bb0-4cb8-b92a-e8f475844d4e)
the FLAC binaries are in:
https://ftp.osuosl.org/pub/xiph/releases/flac/

in command line, writing flac.exe -f "FLACfile" , just writes the values properly, not changing the file timestamp.
can you check it out? and maybe if it suits may you incorporate this method as an option in the program...
add a toggle "write CRC and Encoder" checkbox button, if the user checks it, it will write for all the files the CRC and Encoder fields.

Let me know

**Comment by orbittwz** [comment]
some more information, it does reencode... fine by me.
why? because I saw the timestamp doesn't change.
it doesn't change by default (from the FLAC help):
      --preserve-modtime       Output files keep timestamp of input (default)

the timestamp is being kept by default.
so it does reencode it.
hope you can insert my suggestion above.
thanks

**Comment by orbittwz** [comment]
this is not OK because clearly something is not right... FLAC files must contain how they were encoded and what are their CRC hash after being encoded...
so while the thread is closed, DJ... I beg you to further test why the downloaded streams are being "stripped" somehow...
maybe it's the API of yours you are using, maybe the program... something is not right :( :(
let me know...

**Comment by DJDoubleD** [comment]
> this is not OK because clearly something is not right... FLAC files must contain how they were encoded and what are their CRC hash after being encoded...
Please further test why the downloaded streams are being "stripped" somehow...
maybe it's the API of yours you are using, maybe the program... something is messing with the files.
let me know...

This really has nothing to do with my use of the API nor the way I save the downloaded files.
The simple fact is that the source files on the Qobuz servers **_do not_** contain this data.

Feel free to try any of the other Qobuz downloaders out there ([streamrip](https://github.com/nathom/streamrip), [qobuz-dl](https://github.com/vitiko98/qobuz-dl), [OrpheusDL](https://github.com/OrfiDev/orpheusdl-qobuz), ...).
You will find that they will also produce the exact same audio stream with `MD5 of the unencoded content : 00000000000000000000000000000000` (so no CRC) and possibly something like `Writing library             : Mutagen 1.46.0`, which is only the result of writing the **metadata** with the [mutagen library](https://mutagen.readthedocs.io/en/latest/) (which obviously isn't the original encoding library).

There really isn't anything to be done here, I'm afraid.
